# **Probability 1 Project**

$
\textcolor{gray}{Student}: \textit{Bardia Barghi}
$

$
\textcolor{gray}{Number}: \textit{610303118}
$

### **Part 1**: Calculating Posterior Probabilities With *Discrete Priors*



We first calculate all the *Likelihoods* for different $\theta$ values.

$\theta$ values are represented by a set of 10 possible values ranging from 0.1 to 1 in 0.1 steps: $\theta$ $\in$ $\{0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1\}$

With the Prior belief being: $\space$ $\pi_{Ghader}(\theta)$ = $\{0.05, 0.05, 0.1, 0.1, 0.2, 0.2, 0.1, 0.1, 0.05, 0.05\}$

The *Likelihood* for each value of  $\theta$  (with values `n = 20` and `k = 12`) has a *Binomial* Distribution as written below:

$$
\textcolor{cyan}{Likelihood}(\theta) = \displaystyle\binom{20}{12} \cdot (\theta) ^ {12} \cdot (1 - \theta) ^ 8
$$

By having the *likelihood* of each  $\theta$, We can also calculate the $\textit{Unnormalized Posterior Probability}$ of $\theta$ simply by multiplying the *Likelihood* and the *Prior Probability*: 

$$
\textcolor{cyan}{Unnormalized \; Posterior}(\theta) = P(X ∣ \theta) ⋅ P(\theta) = \text{Likelihood} \times \text{Prior}
$$

The $\textit{Noramlized Posterior Probability}$ can be derived from dividing the *Unnormalized Posterior* by the summation of *Unnormalized Posterior* for every $\theta$ $\in$ $\pi_{Ghader}$ $\:$ :

$$
\textcolor{cyan}{Normalized \; Posterior}(\theta) = \frac{Unnormalized \; Posterior (\theta)}{\sum_{\theta^`}P(X ∣ \theta^`) ⋅ P(\theta^`)}
$$

We then create these sets one by one: 

In [1]:
from scipy.special import comb

theta_values = [0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1]
prior_probability = [0.05, 0.05, 0.1, 0.1, 0.2, 0.2, 0.1, 0.1, 0.05, 0.05]
likelihood = [float(format((comb(20, 12) * (i ** 12) * ((1 - i) ** (20 - 12))), 'f')) for i in theta_values]
unnormalized_posterior = [prior_probability[i] * likelihood[j] for i in range(10) for j in range(10) if i == j]
normalized_posterior = [i / sum(unnormalized_posterior) for i in unnormalized_posterior]

And finally, We build our table with the data at hand:

![Table Image](https://i.ibb.co/b5WDFB2j/Table.png)

And here's the code snippet for printing out the table:

In [2]:
from scipy.special import comb
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

N = 100000

theta_values = [0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1]
prior_probability = [0.05, 0.05, 0.1, 0.1, 0.2, 0.2, 0.1, 0.1, 0.05, 0.05]
likelihood = [float(format((comb(20, 12) * (i ** 12) * ((1 - i) ** (20 - 12))), 'f')) for i in theta_values]
unnormalized_posterior = [prior_probability[i] * likelihood[j] for i in range(10) for j in range(10) if i == j]
normalized_posterior = [i / sum(unnormalized_posterior) for i in unnormalized_posterior]

data = {
    "Theta": theta_values,
    "Prior": prior_probability,
    "Likelihood": likelihood,
    "Unnormalized Posterior": unnormalized_posterior,
    "Normalized Posterior": normalized_posterior
}

df = pd.DataFrame(data)
print(df)


   Theta  Prior  Likelihood  Unnormalized Posterior  Normalized Posterior
0    0.1   0.05    0.000000                0.000000              0.000000
1    0.2   0.05    0.000087                0.000004              0.000056
2    0.3   0.10    0.003859                0.000386              0.004974
3    0.4   0.10    0.035497                0.003550              0.045754
4    0.5   0.20    0.120134                0.024027              0.309697
5    0.6   0.20    0.179706                0.035941              0.463270
6    0.7   0.10    0.114397                0.011440              0.147454
7    0.8   0.10    0.022161                0.002216              0.028565
8    0.9   0.05    0.000356                0.000018              0.000229
9    1.0   0.05    0.000000                0.000000              0.000000


And for the final section of this part, we answer some questions:

1. $\text{What is the most likely value of θ before observing the data? (i.e., the value with the highest prior probability)}$

$\textcolor{orange}{Answer}$: `θ = 0.5` and `θ = 0.6`

2. $\text{What is the most likely value of θ after observing the data? (i.e., the value with the highest posterior probability)}$

$\textcolor{orange}{Answer}$: `θ = 0.6`

3. $\text{What is the posterior probability that θ > 0.5?}$

$\textcolor{orange}{Answer}$: NP(θ = 0.6) + NP(θ = 0.7) + NP(θ = 0.8) + NP(θ = 0.9) + NP(θ = 1) = 0.463270 + 0.147454 + 0.028565 + 0.000229 + 0.000000 = `0.639518`

### **Part 2**: Calculating Posterior Probabilities With *Continuous Priors*


In this part, $\theta$ can take on values within the (0, 1) interval.

A common approach in Bayesian statistics, especially when no specific prior knowledge is available, is to use a **Uniform Prior** over the interval (0,1):

$$
\theta{\quad \sim \quad \text{Uniform(0, 1)}} \quad \to \quad \theta \quad \sim \quad  \text{Beta\text{(1, 1)}}
$$

Again, We observe data from `n = 20` individuals, and `s = 12` of them respond that **Monday** is their favourite day to buy breakfast, and we use this data to update our *Prior* and obtain *Posterior*:

$$
\text{Posterior}(\theta) \quad \sim \quad \text{Likelihood}(\theta)
$$

Precisely:

$$
\text{Posterior}(\theta) \quad = \quad \frac{Likelihood(\theta)}{\displaystyle \int_0^1 Likelihood(\theta) \: \mathrm{d}\theta} \quad = \quad \frac{\displaystyle\binom{20}{12} \cdot \theta ^ {12} \cdot (1 - \theta) ^ 8
}{\displaystyle \int_0^1 \displaystyle\binom{20}{12} \cdot \theta ^ {12} \cdot (1 - \theta) ^ 8 \: \mathrm{d}\theta}
$$

$$
\text{Beta(a, b)} \quad = \quad \displaystyle{\int_0^1 x^{a - 1} \cdot (1 - x)^{b - 1}} \: \mathrm{d}x \quad = \quad \frac{\Gamma(a) \cdot \Gamma(b)}{\Gamma(a + b)}
$$

By simplifying the *Posterior* we get:

$$
\text{Posterior}(\theta) \quad = \quad \frac{\theta^{12} \cdot (1 - \theta)^{8}}{\frac{\Gamma(13) \cdot \Gamma(9)}{\Gamma(21)}} \quad = \quad \frac{\Gamma(21)}{\Gamma(13) \cdot \Gamma(9)} \cdot \theta^{12} \cdot (1 - \theta)^{8} 
$$